pyteal/ast/frame.py

```diff
@@ -1,27 +1,172 @@
-from typing import TYPE_CHECKING
+from itertools import groupby
+from typing import TYPE_CHECKING, Optional
 
 from pyteal.ast.expr import Expr
+from pyteal.ast.int import Int
+from pyteal.ast.bytes import Bytes
+from pyteal.ast.abstractvar import AbstractVar
 from pyteal.types import TealType, require_type
-from pyteal.errors import TealInputError, verifyProgramVersion
+from pyteal.errors import TealInputError, TealInternalError, verifyProgramVersion
 from pyteal.ir import TealBlock, TealSimpleBlock, TealOp, Op
 
 if TYPE_CHECKING:
     from pyteal.compiler import CompileOptions
 
 
+class LocalTypeSegment(Expr):
+    """An expression that allocates stack spaces for local variable.
+
+    This class is intentionally hidden because it's too basic to directly expose.
+    This is only used in ProtoStackLayout internally.
+    """
+
+    def __init__(self, local_type: TealType, count: int):
+        super().__init__()
+        self.local_type = local_type
+        self.count = count
+        self.auto_instance: Expr
+
+        if self.count <= 0:
+            raise TealInternalError(
+                "LocalTypeSegment initialization error: segment length must be strictly greater than 0."
+            )
+        match self.local_type:
+            case TealType.uint64 | TealType.anytype:
+                self.auto_instance = Int(0)
+            case TealType.bytes:
+                self.auto_instance = Bytes("")
+            case TealType.none:
+                raise TealInternalError(
+                    "Local variable in subroutine initialization must be typed."
+                )
+
+    def __teal__(self, options: "CompileOptions") -> tuple[TealBlock, TealSimpleBlock]:
+        if self.count == 1:
+            inst_srt, inst_end = self.auto_instance.__teal__(options)
+            return inst_srt, inst_end
+        else:
+            dupn_srt, dupn_end = DupN(self.auto_instance, self.count).__teal__(options)
```

Comment:
while trying this out I found that DupN adds an extra element to the stack, since we've already got one from the initial instance we should use `self.count - 1` I think, but please double check me.

prior to change:
```
127 |69  |callsub label9                    |        |[]
203 |123 |proto 0 0                         |        |[]
206 |124 |intc_0 // 0                       |        |[]
207 |125 |dupn 3                            |        |[0]
209 |126 |txna ApplicationArgs 1            |        |[0, 0, 0, 0]
212 |127 |btoi                              |        |[0, 0, 0, 0, 0x000000000000000...
213 |128 |frame_bury 1                      |        |[0, 0, 0, 0, 2]
215 |129 |txna ApplicationArgs 2            |        |[0, 2, 0, 0]
218 |130 |btoi                              |        |[0, 2, 0, 0, 0x000000000000000...
219 |131 |frame_bury 2                      |        |[0, 2, 0, 0, 4]
221 |132 |callsub label12                   |        |[0, 2, 4, 0]
```
After:
```
127 |69  |callsub label9                    |        |[]
203 |123 |proto 0 0                         |        |[]
206 |124 |intc_0 // 0                       |        |[]
207 |125 |dupn 2                            |        |[0]
209 |126 |txna ApplicationArgs 1            |        |[0, 0, 0]
212 |127 |btoi                              |        |[0, 0, 0, 0x0000000000000002]
213 |128 |frame_bury 1                      |        |[0, 0, 0, 2]
215 |129 |txna ApplicationArgs 2            |        |[0, 2, 0]
218 |130 |btoi                              |        |[0, 2, 0, 0x0000000000000004]
219 |131 |frame_bury 2                      |        |[0, 2, 0, 4]
221 |132 |callsub label12                   |        |[0, 2, 4]
```


```suggestion
            dupn_srt, dupn_end = DupN(self.auto_instance, self.count - 1).__teal__(options)
```

Reply:
right, that makes sense, I was unaware of the dupn doesn't consume the initial one on stack, and your finding is correct.